Remotion/Data/DomainObjects/Mapping/InlineTypeDefinitionVisitor.cs

```diff
@@ -0,0 +1,118 @@
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+        _classDefinitionHandler = classDefinitionHandler;
+        _interfaceDefinitionHandler = interfaceDefinitionHandler;
+      }
+
+      /// <inheritdoc />
+      public T? VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        return _classDefinitionHandler(this, classDefinition);
+      }
+
+      public T? VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        ArgumentUtility.CheckNotNull("interfaceDefinition", interfaceDefinition);
+
+        return _interfaceDefinitionHandler(this, interfaceDefinition);
+      }
+    }
+
+    private class TypeDefinitionVisitor : ITypeDefinitionVisitor
+    {
+      private readonly Action<ITypeDefinitionVisitor, ClassDefinition> _classDefinitionHandler;
+      private readonly Action<ITypeDefinitionVisitor, InterfaceDefinition> _interfaceDefinitionHandler;
+
+      public TypeDefinitionVisitor (
+          Action<ITypeDefinitionVisitor, ClassDefinition> classDefinitionHandler,
+          Action<ITypeDefinitionVisitor, InterfaceDefinition> interfaceDefinitionHandler)
+      {
+        ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+        _classDefinitionHandler = classDefinitionHandler;
+        _interfaceDefinitionHandler = interfaceDefinitionHandler;
+      }
+
+      /// <inheritdoc />
+      public void VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        _classDefinitionHandler(this, classDefinition);
+      }
+
+      /// <inheritdoc />
+      public void VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        ArgumentUtility.CheckNotNull("interfaceDefinition", interfaceDefinition);
+
+        _interfaceDefinitionHandler(this, interfaceDefinition);
+      }
+    }
+
+    public static T? Visit<T> (
+        TypeDefinition typeDefinition,
+        Func<ITypeDefinitionVisitor<T>, ClassDefinition, T> classDefinitionHandler,
+        Func<ITypeDefinitionVisitor<T>, InterfaceDefinition, T> interfaceDefinitionHandler)
+    {
+      ArgumentUtility.CheckNotNull(nameof(typeDefinition), typeDefinition);
+      ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+      ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+      var visitor = new TypeDefinitionVisitor<T>(classDefinitionHandler, interfaceDefinitionHandler);
+      return typeDefinition.Accept(visitor);
+    }
+
+    public static void Visit (
+        TypeDefinition typeDefinition,
+        Action<ITypeDefinitionVisitor, ClassDefinition> classDefinitionHandler,
```

Comment:
Why do we pass the visitor into the handler? The other inline-API doesn't do this and it feels kind of strange?

Reply:
The inline walker already calls itself to walk the hierarchy, as such, it is not necessary to manually call the visitor. If something like that is necessary one should create a class derived from the base walker.

In the case of the inline visitor it is a much more common use case to continue calling the visitor. This is not possible without the argument as the visitor is never available from the outside.

I had a use case for this, but I fear that this was refactored as I cannot find it anymore. While I think that this solution has better usability, I have no problem removing the parameter.

Reply:
Then I opt for dropping this to keep the API more straight forward until the need actually arises.